widget/list.go

```diff
@@ -0,0 +1,449 @@
+	}
+}
+
+func (l *listRenderer) setupListItem(item fyne.CanvasObject, index int) {
+	previousIndicator := item.(*listItem).selected
+	if index != l.list.selectedIndex {
+		item.(*listItem).selected = false
+	} else {
+		item.(*listItem).selected = true
+		l.list.selectedItem = item.(*listItem)
+	}
+	if previousIndicator != item.(*listItem).selected {
+		item.Refresh()
+	}
+	l.list.UpdateItem(index, item.(*listItem).child)
+	item.(*listItem).onTapped = func() {
+		if l.list.selectedItem != nil && l.list.selectedIndex >= l.firstItemIndex && l.list.selectedIndex <= l.lastItemIndex {
+			l.list.selectedItem.selected = false
+			l.list.selectedItem.Refresh()
+		}
+		l.list.selectedItem = item.(*listItem)
+		l.list.selectedIndex = index
+		l.list.selectedItem.selected = true
+		l.list.selectedItem.Refresh()
+		if l.list.OnItemSelected != nil {
+			l.list.OnItemSelected(index, item.(*listItem).child)
+		}
+	}
+}
+
+type listItem struct {
+	BaseWidget
+
+	onTapped          func()
+	statusIndicator   *canvas.Rectangle
+	child             fyne.CanvasObject
+	divider           *canvas.Rectangle
+	hovered, selected bool
+}
+
+func newListItem(child fyne.CanvasObject, tapped func()) *listItem {
+	li := &listItem{
+		child:    child,
+		onTapped: tapped,
+	}
+
+	li.ExtendBaseWidget(li)
+	return li
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (li *listItem) CreateRenderer() fyne.WidgetRenderer {
+	li.ExtendBaseWidget(li)
+
+	li.statusIndicator = canvas.NewRectangle(theme.BackgroundColor())
+	li.divider = canvas.NewRectangle(theme.ShadowColor())
+
+	objects := []fyne.CanvasObject{li.statusIndicator, li.child, li.divider}
+
+	return &listItemRenderer{widget.NewBaseRenderer(objects), li}
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (li *listItem) MinSize() fyne.Size {
+	li.ExtendBaseWidget(li)
+	return li.BaseWidget.MinSize()
+}
+
+// MouseIn is called when a desktop pointer enters the widget
+func (li *listItem) MouseIn(*desktop.MouseEvent) {
+	li.hovered = true
+	li.Refresh()
+}
+
+// MouseMoved is called when a desktop pointer hovers over the widget
+func (li *listItem) MouseMoved(*desktop.MouseEvent) {
+}
+
+// MouseOut is called when a desktop pointer exits the widget
+func (li *listItem) MouseOut() {
+	li.hovered = false
+	li.Refresh()
+}
+
+// Tapped is called when a pointer tapped event is captured and triggers any tap handler
+func (li *listItem) Tapped(*fyne.PointEvent) {
+	if li.onTapped != nil {
+		li.selected = true
+		li.Refresh()
+		li.onTapped()
+	}
+}
+
+type listItemRenderer struct {
+	widget.BaseRenderer
+
+	item *listItem
+}
+
+// MinSize calculates the minimum size of a listItem.
+// This is based on the size of the status indicator and the size of the child object.
+func (li *listItemRenderer) MinSize() (size fyne.Size) {
+	size = fyne.NewSize((theme.Padding()*2)+li.item.child.Size().Width,
+		li.item.child.Size().Height+theme.Padding()*2)
```

Comment:
Is it possible that li.item.child.Size().IsZero(), perhaps this should be li.item.child.MinSize()?

Reply:
I'm not sure I follow... You are suggestion if the item min size is zero that we shouldn't add the padding in and just make it 0?  I don't think that would work as that would not leave room for the indicator or divider in the list item.

Reply:
Not quite, I'm saying li.item.child.Size() can return 0x0

Reply:
Okay resolved